google-cloud-pubsub/src/main/java/com/google/cloud/pubsub/v1/StreamingSubscriberConnection.java

```diff
@@ -319,52 +378,354 @@ private boolean isAlive() {
     return state == State.RUNNING || state == State.STARTING;
   }
 
-  @Override
-  public void sendAckOperations(
-      List<String> acksToSend, List<PendingModifyAckDeadline> ackDeadlineExtensions) {
-    ApiFutureCallback<Empty> loggingCallback =
-        new ApiFutureCallback<Empty>() {
-          @Override
-          public void onSuccess(Empty empty) {
-            ackOperationsWaiter.incrementPendingCount(-1);
-          }
+  public void setResponseOutstandingMessages(AckResponse ackResponse) {
+    // We will close the futures with ackResponse - if there are multiple references to the same
+    // future they will be handled appropriately
+    logger.log(
+        Level.WARNING, "Setting response: {0} on outstanding messages", pendingRequests.size());
+    for (AckRequestData ackRequestData : pendingRequests) {
+      ackRequestData.setResponse(ackResponse, false);
+    }
 
-          @Override
-          public void onFailure(Throwable t) {
-            ackOperationsWaiter.incrementPendingCount(-1);
-            Level level = isAlive() ? Level.WARNING : Level.FINER;
-            logger.log(level, "failed to send operations", t);
-          }
-        };
+    // Clear our pending requests
+    pendingRequests.clear();
+  }
 
-    int pendingOperations = 0;
-    for (PendingModifyAckDeadline modack : ackDeadlineExtensions) {
-      for (List<String> idChunk : Lists.partition(modack.ackIds, MAX_PER_REQUEST_CHANGES)) {
-        ApiFuture<Empty> future =
-            stub.modifyAckDeadlineCallable()
-                .futureCall(
-                    ModifyAckDeadlineRequest.newBuilder()
-                        .setSubscription(subscription)
-                        .addAllAckIds(idChunk)
-                        .setAckDeadlineSeconds(modack.deadlineExtensionSeconds)
-                        .build());
-        ApiFutures.addCallback(future, loggingCallback, directExecutor());
-        pendingOperations++;
+  private void setFailureFutureOutstandingMessages(Throwable t) {
+    if (shouldSetMessageFuture()) {
```

Comment:
Isn't this a duplicate check? We wouldn't have a message future in the first place if this was false.

If so, we can remove this method, and we no longer have to store the messageReceiverWithResponse in StreamingSubscriberConnection

Reply:
I had this as a short-circuit, but I think it will be more straightforward to remove it